### ISSUE TITLE
Synchronized cursors between the 2 maps

### ISSUE THREAD

**Issue by sylvain-m** [turban/Leaflet.Sync#27]
Hello,

I'm no expert in Github, and this is not really an issue but a pull request (I didn't succeed to pull request for this script).
I managed to integrate in a page 2 synchronized maps : thank you for this great script !!
However, I miss a very useful feature: synchronized cursors between the 2 maps. These cursors allow you to accurately compare the location pointed to by the mouse. This is very useful for diachronic cards (comparing 2 dates).

Is this already possible, or is this an improvement to be made to script?
Thank you for your help and congratulations for your great work!

Sylvain M.

**Comment by sylvain-m** [comment]
Here is an example of what i would like, based on OpenLayers 2 :
[http://geobretagne.fr/sviewer/dual.html](http://geobretagne.fr/sviewer/dual.html)

**Comment by jieter** [comment]
@sylvain-m nice suggestion. Since you are not providing a working solution, that makes this an issue and not a pull request (you requesting me to pull some code from your repository into mine).

Anyway, this might be an interesting feature to have. It's not possible in the current state of the script, but I guess it wouldn't be that hard to add it. Would you like to try to add it? I'm willing to help you in the process.

**Comment by sylvain-m** [comment]
> Would you like to try to add it?
For sure ! But i'm not a real webmapper :-$

I adapt the codes I find here and there, but I'm not able to create them myself!
The URL of my page is not yet public (there are still too many loose ends), but if I can send it privately, it would be very useful for our project.
I understand that you agree to help me implement this function on my page? (I do not speak much English, so I'm afraid to misinterpret your proposal)

**Comment by jieter** [comment]
I would be willing to guide you while adding cursor sync to Leaflet.Sync, not that I'll do all the coding for you.

**Comment by baipi** [comment]
You can easily add it, could be something like that :+1: 
`
cursor1 = L.circleMarker([0,0], {radius:25, fillOpacity: 0.1, color: '#FFF48D', fillColor: '#FFFFFF'}).addTo(map1);
cursor2 = L.circleMarker([0,0], {radius:25, fillOpacity: 0.1, color: '#FFF48D', fillColor: '#FFFFFF'}).addTo(map2);
 map1.on('mousemove', function (e) {
   cursor1.setLatLng(e.latlng);
   cursor2.cursorr.setLatLng(e.latlng);
 });
map2.on('mousemove', function (e) {
   cursor1.setLatLng(e.latlng);
   cursor2.cursorr.setLatLng(e.latlng);
 });`

**Comment by sylvain-m** [comment]
Great !!!
How quickly Baipi!
I managed to integrate your code, and it works (almost) perfectly!
There are just two small problems:
- When you mouse on the map 1, the cursor is not on the map 2 (if the mouse is on map 2, it is good on the map 1)
- Cursor of map 2 is a hand, and not the circle.
But a huge thank you for these very useful lines !!!

**Comment by sylvain-m** [comment]
I have solve the first point with this code
```
     cursor1 = L.circleMarker([0,0], {radius:25, fillOpacity: 0.1, color: '#FFF48D', fillColor: '#FFFFFF'}).addTo(map1);
     cursor2 = L.circleMarker([0,0], {radius:25, fillOpacity: 0.1, color: '#FFF48D', fillColor: '#FFFFFF'}).addTo(map2);
     map1.on('mousemove', function (e) { cursor2.setLatLng(e.latlng); cursor1.cursorr.setLatLng(e.latlng); });
     map2.on('mousemove', function (e) { cursor1.setLatLng(e.latlng); cursor2.cursorr.setLatLng(e.latlng); });
```
Second point is still here : cursor of the "real" mouse on the screen is a hand and not a circle

**Comment by sylvain-m** [comment]
Sorry, i didn't saw you modified the code between my messages !
I try !

**Comment by jieter** [comment]
The next step: integrate this cleanly into leaflet.sync code... ;)

**Comment by baipi** [comment]
Had a CSS class to your circleMarker with the property "cursor" at "none".


JS: 
```
cursor1 = L.circleMarker([0,0], {radius:25, fillOpacity: 0.1, color: '#FFF48D', fillColor: '#FFFFFF', className: 'noCursor'}).addTo(map1);
```

CSS:
```
.noCursor{
  cursor:none;
}
```

**Comment by sylvain-m** [comment]
IS JS for cursor1 only, or for both cursor1 and cursor2 ?
(Sorry, I'm really not good in javascript!)

**Comment by baipi** [comment]
No problem :)
Do it for both, it should works :)

**Comment by sylvain-m** [comment]
I tried, but did not succeed...
Here is the URL of test web page : [http://www.parc-naturel-normandie-maine.fr/upload/gmap/observatoire/wmts/bdbocage_dual_cursorsync.html](http://www.parc-naturel-normandie-maine.fr/upload/gmap/observatoire/wmts/bdbocage_dual_cursorsync.html)
(il will supress it after you saw it)

**Comment by jieter** [comment]
hmm, you problem is the circlemarker on the map you are moving over doesn't get updated. 

```
map1.on('mousemove', function (e) { cursor2.setLatLng(e.latlng); cursor1.cursorr.setLatLng(e.latlng); });
map2.on('mousemove', function (e) { cursor1.setLatLng(e.latlng); cursor2.cursorr.setLatLng(e.latlng); });
```

You should remove `.cursorr` twice.

**Comment by baipi** [comment]
```
cursor1 = L.circleMarker([0,0], {radius:25, fillOpacity: 0.1, color: '#FFF48D', fillColor: '#FFFFFF', className: 'noCursor'}).addTo(map1);
cursor2 = L.circleMarker([0,0], {radius:25, fillOpacity: 0.1, color: '#FFF48D', fillColor: '#FFFFFF', className: 'noCursor'}).addTo(map2);
map1.on('mousemove', function (e) { 
    cursor2.setLatLng(e.latlng); 
    cursor1.setLatLng(e.latlng); // Not cursor1.cursorr.setLatLng(e.latlng);
});

map2.on('mousemove', function (e) { 
    cursor1.setLatLng(e.latlng); 
    cursor2.setLatLng(e.latlng); //  Not cursor2.cursorr.setLatLng(e.latlng); 
 });
```

**Comment by baipi** [comment]
@jieter is faster :)

Je vois que ça marche !

**Comment by sylvain-m** [comment]
Great !!
Thank you both !
It is on: how effective and fast!
Another good example of the effectiveness of Open Source, I will not fail to promote!

**Comment by jieter** [comment]
@sylvain-m 

This is a rough implemenation in Leaflet.sync: https://github.com/turban/Leaflet.Sync/blob/sync-cursor/L.Map.Sync.js#L35-L52 It would be nice if someone is willing to complete that. @baipi ?

**Comment by baipi** [comment]
@jieter I looking on it as soon as possible :)

**Comment by jieter** [comment]
nice. Some considerations:
-  adding the listeners might be better at home in `_initSync()`, which needs to get the sync options to decide if it needs to sync cursors. 
- Things need to be cleaned up properly in 'unsync'.

**Comment by RalucaNicola** [comment]
I tried to solve the unsync part (as an exercise to practice my beginner JavaScript skills), but I am not sure if it's correct. Could any of you guys have a look at my code? 

https://github.com/RalucaNicola/Leaflet.Sync/blob/new-sync-cursor/L.Map.Sync.js 
In the unsync method I just remove the cursors corresponding to the map. Not sure however that this is the "proper" way: @jieter You mentioned that the event listeners should be removed, but if there is more than one cursor in the array, removing the event listener will stop the other cursors from working. 

And one question: why do we need this mouseout event? If the mouse leaves the map, the cursor leaves the map too, which is ok...isn't it?

**Comment by jieter** [comment]
@RalucaNicola can you open a pull request, that will make it easier to review your patch.

**Comment by RalucaNicola** [comment]
I would, but I was fighting some detached HEAD last night, so the only way that worked for me to add my code was to make a new branch (new-sync-cursor). Should I open a pull request anyway?
Basically I just added these 3 lines to remove the cursors: https://github.com/RalucaNicola/Leaflet.Sync/commit/99e13743745fca4c49e0e6aa8cf0e6183626a598

**Comment by jieter** [comment]
@RalucaNicola yeah, open a PR anyway. That will trigger the unit tests too. Would also be nice if you could add some unit tests to test this cursor syncing, or at least the proper cleanup after unsyncing.

Don't be afraid to propose incomplete changes, we can discuss and improve as we go.

**Comment by RalucaNicola** [comment]
@jieter ok, I hope I did it right, this was my first pull request https://github.com/turban/Leaflet.Sync/pull/29 it shows that some check failed though...I also tested if it works in the examples. Thank you for reviewing it!

**Comment by jieter** [comment]
closed via #30 and 0a9261dffa23dd52899ad452d5892242b2ed9a32
